auction-server/src/opportunity/api.rs

```diff
@@ -102,21 +107,105 @@ pub struct OpportunityParamsV1 {
     #[serde(with = "crate::serde::u256")]
     pub target_call_value: U256,
 
-    pub sell_tokens: Vec<TokenAmount>,
-    pub buy_tokens:  Vec<TokenAmount>,
+    pub sell_tokens: Vec<TokenAmountEvm>,
+    pub buy_tokens:  Vec<TokenAmountEvm>,
 }
 
 #[derive(Serialize, Deserialize, ToSchema, Clone, PartialEq, Debug)]
 #[serde(tag = "version")]
+pub enum OpportunityParamsEvm {
+    #[serde(rename = "v1")]
+    V1(OpportunityParamsV1Evm),
+}
+
+#[derive(Serialize, Deserialize, ToSchema, Clone, PartialEq, Debug)]
+#[serde(untagged)]
 pub enum OpportunityParams {
+    Evm(OpportunityParamsEvm),
+    Svm(OpportunityParamsSvm),
+}
+
+#[serde_as]
+#[derive(Serialize, Deserialize, ToSchema, Clone, PartialEq, Debug)]
+pub struct TokenAmountSvm {
+    /// Token contract address
+    #[schema(example = "DUcTi3rDyS5QEmZ4BNRBejtArmDCWaPYGfN44vBJXKL5", value_type = String)]
+    #[serde_as(as = "DisplayFromStr")]
+    pub token:  Pubkey,
+    /// Token amount
+    #[schema(example = 1000)]
+    pub amount: u64,
+}
+
+// TODO descibe it
+#[serde_as]
+#[derive(Serialize, Deserialize, ToSchema, Clone, PartialEq, Debug)]
+pub struct OpportunityParamsV1KaminoSvm {
+    #[schema(example = "DUcTi3rDyS5QEmZ4BNRBejtArmDCWaPYGfN44vBJXKL5", value_type = String)]
+    #[serde_as(as = "DisplayFromStr")]
+    pub permission: Pubkey,
+    #[schema(example = "DUcTi3rDyS5QEmZ4BNRBejtArmDCWaPYGfN44vBJXKL5", value_type = String)]
+    #[serde_as(as = "DisplayFromStr")]
+    pub router:     Pubkey,
+    #[schema(example = "solana", value_type = String)]
+    pub chain_id:   ChainId,
+
+    pub sell_tokens: Vec<TokenAmountSvm>,
+    pub buy_tokens:  Vec<TokenAmountSvm>,
+}
+
+#[derive(Serialize, Deserialize, ToSchema, Clone, PartialEq, Debug)]
+#[serde(tag = "client")]
+pub enum OpportunityParamsV1Svm {
+    #[serde(rename = "kamino")]
+    Kamino(OpportunityParamsV1KaminoSvm),
+}
+
+#[derive(Serialize, Deserialize, ToSchema, Clone, PartialEq, Debug)]
+#[serde(tag = "version")]
+pub enum OpportunityParamsSvm {
     #[serde(rename = "v1")]
-    V1(OpportunityParamsV1),
+    V1(OpportunityParamsV1Svm),
+}
+
+/// Similar to OpportunityParams, but with the opportunity id included.
+#[derive(Serialize, Deserialize, ToSchema, Clone, ToResponse)]
+pub struct OpportunityParamsWithMetadataEvm {
```

Comment:
can this become a generic? Since it's getting exposed to apis and axum through two concrete types from the OpportunityParamsWithMetadata enum, maybe it will work?

Reply:
I change types a bit, but didnt make them generic as we need these types for open api generation.